Comments/Services/OrganisationService.cs

```diff
@@ -0,0 +1,185 @@
+
+			return new OrganisationCode(organisationAuthorisation, currentUsersOrganisationName);
+        }
+
+		/// <summary>
+		/// this outputs a collation code like this: "[4 numbers][space][4 numbers][space][4 numbers]"
+		///
+		/// the spaces are stripped before saving in the database, as the database column is nvarchar(12)
+		/// </summary>
+		/// <returns></returns>
+		private string GenerateCollationCode()
+        {
+	        var random = new Random();
+	        var firstPart = random.Next(100000, 999999); //preventing a leading zero, which a user might not type if they consider it a number.
+	        var secondPart = random.Next(0, 999999);
+	        var collationCode = $"{firstPart:0000 00}{secondPart:00 0000}";
+	        return collationCode;
+        }
+
+		/// <summary>
+		/// Checks a collation code is valid for a given consultation
+		/// </summary>
+		/// <param name="collationCode"></param>
+		/// <param name="consultationId"></param>
+		/// <returns>Returns null if the collation code is not valid</returns>
+		public async Task<OrganisationCode> CheckValidCodeForConsultation(string collationCode, int consultationId)
+		{
+			var genericCodeErrorMessage = "Incorrect code. Verify the code with your organisation's commenting lead and try again.";
+
+			var organisationAuthorisation = _context.GetOrganisationAuthorisationByCollationCode(collationCode);
+			if (organisationAuthorisation == null)
+				throw new ApplicationException(genericCodeErrorMessage); //old message: "Collation code not found"
+
+			var sourceURI = ConsultationsUri.CreateConsultationURI(consultationId);
+			if (!organisationAuthorisation.Location.SourceURI.Equals(sourceURI, StringComparison.OrdinalIgnoreCase))
+				throw new ApplicationException(genericCodeErrorMessage); //old message: "The supplied code is for a different consultation."
+
+			var hasOrganisationSubmittedForThisConsultation = HasOrganisationSubmittedForConsultation(organisationAuthorisation.OrganisationId, sourceURI);
+			if (hasOrganisationSubmittedForThisConsultation)
+				throw new ApplicationException("Your organisation has already responded to this consultation. Contact your organisation's commenting lead for further information.");
+
+			//var machineToMachineAccessToken =	await _apiTokenService.GetAccessToken(AppSettings.AuthenticationConfig.GetAuthConfiguration()); 
+			//var httpClientWithPooledMessageHandler = _httpClientFactory.CreateClient();
+
+			//var organisations = await _apiService.GetOrganisations(new List<int> {organisationAuthorisation.OrganisationId}, machineToMachineAccessToken,	httpClientWithPooledMessageHandler);
+
+			var organisations = new List<Organisation> {new Organisation(1, "Faculty of Pain Medicine of the Royal College of Anaesthetists", false)}; //TODO: remove. this is only here since we're not currently caching the token. Once idam's been updated with m2m token caching we can restore the above code and remove this.
```

Comment:
each time that code gets called it consumes 1 machine-to-machine (client_credentials) token. potentially it could get called multiple times when the user types in the code (though there is some debouncing).

we've got a fixed quota of them per month. so it needs caching.

Jose's written the code to cache the token in redis. i'm just waiting on him committing and merging it before uncommenting this.